### DISCUSSION TITLE
NotifierProvider does not provide ref

### DISCUSSION THREAD

**Discussion by WingCH** [rrousselGit/riverpod#2084] [Q&A]
I am currently learning Riverpod and I've noticed that the NotifierProvider does not provide a ref, while the StateNotifierProvider does. I am wondering if there are any design considerations behind this choice.  

I would like to pass a ref to Notifier, since I am trying to put a dependency service in the Notifier.  

Thank you for your help.

```dart
// NotifierProvider
final todosProvider = NotifierProvider<TodosNotifier, List<Todo>>(() {
  return TodosNotifier();
})

// StateNotifierProvider
final todosProvider = StateNotifierProvider<TodosNotifier, List<Todo>>((ref) {
  return TodosNotifier();
});
```

**Comment by WingCH** [comment]
I have found the relevant description in the document, it would be better if it can be added to the tab of `(Async) NotifierProvider`

https://docs-v2.riverpod.dev/docs/concepts/provider_lifecycles#alive

> Typically when constructing a [StateNotifier](https://pub.dev/documentation/state_notifier/latest/state_notifier/StateNotifier-class.html) or [ChangeNotifier](https://api.flutter.dev/flutter/foundation/ChangeNotifier-class.html) class you should pass in the ref to allow the [Notifier](https://pub.dev/documentation/riverpod/latest/riverpod/Notifier-class.html) to obtain the current value of dependencies as needed. By using the new Notifier and [AsyncNotifier](https://pub.dev/documentation/riverpod/latest/riverpod/AsyncNotifier-class.html) classes from Riverpod 2.0, the ref is already available as an instance member of the class.

**Comment by rrousselGit** [comment]
THe `ref` is in the `Notifier` class itself:

```dart
class MyNotifier extends Notifier<int> {
  int build() {
    ref.watch(...)
  }
}
```

**Reply by nero-angela** [reply]
If `ref` can also be used in `NotifierProvider`, I think it will be possible to take advantage of Riverpod's strengths, which can create more diverse providers of the same class.

I wonder if there are any plans to provide `ref` from `NotifierProvider` in the future and whether to deprecate `StateNotifierProvider` someday.
